### PR TITLE
[ONNX] Fix onnx constant folding

### LINKED ISSUE #101328: [ONNX, TorchScript] Constants in ScriptModule got wrong device during optimization
### 🐛 Describe the bug

My repro:

```
import torch
import torch.nn as nn
import torch._dynamo as dynamo


mod = torch.jit.load('scriptmodule.pt')

input0 = torch.randn([1, 128, 48, 80], device="cuda")
input1 = torch.randn([1, 256, 24, 40], device="cuda")
input2 = torch.randn([1, 512, 12, 20], device="cuda")

# this runs fine
mod(input0, input1, input2)

# this will fail
torch.onnx.export(
    mod,
    [input0, input1, input2],
    "test.onnx",
    input_names=['input_0', 'input_1', 'input_2'],
    output_names=['output_0'],
    opset_version=14,
)
```

stacktrace:
```
/home/yj/pytorch/torch/onnx/utils.py:691: UserWarning: Constant folding in symbolic shape inference fails: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument index in method wrapper_CUDA__index_select) (Triggered internally at /home/yj/pytorch/torch/csrc/jit/passes/onnx/shape_type_inference.cpp:439.)
  _C._jit_pass_onnx_graph_shape_type_inference(
========= Diagnostic Run torch.onnx.export version 2.0.0a0+gitc263bd4 ==========
verbose: False, log level: Level.ERROR
======================= 0 NONE 0 NOTE 0 WARNING 0 ERROR ========================

Traceback (most recent call last):
  File "test2.py", line 25, in <module>
    torch.onnx.export(
  File "/home/yj/pytorch/torch/onnx/utils.py", line 507, in export
    _export(
  File "/home/yj/pytorch/torch/onnx/utils.py", line 1591, in _export
    graph, params_dict, torch_out = _model_to_graph(
  File "/home/yj/pytorch/torch/onnx/utils.py", line 1215, in _model_to_graph
    params_dict = _C._jit_pass_onnx_constant_fold(
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument index in method wrapper_CUDA__index_select)
```

### Versions

PyTorch version: 2.0.0a0+gitc263bd4
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.8.15 (default, Nov 24 2022, 15:19:38)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-69-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070
Nvidia driver version: 520.61.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          24
On-line CPU(s) list:             0-23
Thread(s) per core:              2
Core(s) per socket:              12
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           33
Model name:                      AMD Ryzen 9 5900X 12-Core Processor
Stepping:                        0
Frequency boost:                 enabled
CPU MHz:                         2200.000
CPU max MHz:                     3700.0000
CPU min MHz:                     2200.0000
BogoMIPS:                        7399.70
Virtualization:                  AMD-V
L1d cache:                       384 KiB
L1i cache:                       384 KiB
L2 cache:                        6 MiB
L3 cache:                        64 MiB
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm

cc: @BowenBao @jiafatom

### PR DESCRIPTION
**Pull request by shingjan**
Fixes #101328

Note that this most likely is a bandage solution. We either need to actually fix one of those onnx passes that is causing this decomposition/functionalization issue, or need to special case all onnx op in `runTorchBackendForOnnx` like this one.

### CODE REVIEW

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 1]
```diff
@@ -487,6 +487,15 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
     if (q > 1) {
       return c10::nullopt;
     }
+    // If the device of indices tensor is not the same with it of the input tensor
+    // move it to the device of the input device
+    auto indices_val = node->inputs()[1];
+    if(indices_val->node()->kind() == onnx::Constant) {
+      std::cout << "I am here \n";
```
Comment (vadimkantorov) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp:494]: maybe forgot after debugging

Comment (shingjan) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp]: good catch! It is removed

**Comment by vadimkantorov** [comment]
@shingjan I'm not related to maintainers, just found the dangling `cout <<` :)

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 2]
```diff
@@ -487,6 +487,14 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
     if (q > 1) {
       return c10::nullopt;
     }
+    // If the device of indices tensor is not the same with it of the input
+    // tensor, move it to the device of the input tensor
+    auto indices_val = node->inputs()[1];
+    if (indices_val->node()->kind() == onnx::Constant) {
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp:493]: Is this check necessary?

**Review by BowenBao — COMMENTED** [review]
Thanks for contributing @shingjan ! Agree this is a one-off fix. Ideally we should track the origin of misplaced device and fix things there.

File: torch/csrc/jit/passes/onnx/constant_fold.cpp (hunk 2 above)
Comment (shingjan) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp:493]: Hi @BowenBao thanks for the review! This check is needed as the issue is caused by module `constant`s being retrieved back as `cpu` tensors during constant folding. Therefore in this fix we only convert those constant tensor(s) to input device if 1. they are constant 2. they are on a different device than input tensors (most likely cpu vs cuda)

Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp:493]: Hmm for my own knowledge, what are cases when `indices_val` node kind is not onnx::Constant? I think at this point we are guaranteed `indices` a.k.a `inputTensorValues[1]` as tensor data is available to us. Is there other possibilities for `indices_val` nodekind?

And if the answer is yes, shouldn't we also cast if its device is not the same as `inputTensorValues[0]`'s?

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 3]
```diff
@@ -487,6 +487,14 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
     if (q > 1) {
       return c10::nullopt;
     }
+    // If the device of indices tensor is not the same with it of the input
+    // tensor, move it to the device of the input tensor
+    auto indices_val = node->inputs()[1];
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp:492]: nit: `node->input(1)` is safer call.

File: torch/csrc/jit/passes/onnx/constant_fold.cpp (hunk 2 above)
Comment (shingjan) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp]: That makes sense. It is changed. Please take a look!

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 4]
```diff
@@ -487,6 +487,12 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
     if (q > 1) {
       return c10::nullopt;
     }
+    // If the device of indices tensor is not the same with it of the input
+    // tensor, move it to the device of the input tensor
+    auto indices_val = node->inputs(1);
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp:492]:
```suggestion
    auto indices_val = node->input(1);
```

**Review by BowenBao — APPROVED** [review]

**Comment by shingjan** [comment]
Thanks @BowenBao for the review!

**Comment by aidevmin** [comment]
@shingjan 
I see your pull request here https://github.com/pytorch/pytorch/pull/101329/files. I want to change file `torch/csrc/jit/passes/onnx/constant_fold.cpp` but I can not folder `csrc` under `torch` folder. I am using Pytorch 2.0.1, whichs version are you using?

**Comment by shingjan** [comment]
@aidevmin are you using pre-built pytorch, e.g. from conda or pip? If so you probably need to get and build pytorch from source in order to send a PR.

**Comment by aidevmin** [comment]
@shingjan 
Thanks. I install by using pip. Oh Pytorch does not update in pip package. I am using python3.6, it recommend 3.8 when buiding from source.

**Pull request closed without merging**